### ISSUE TITLE
Clarity of error message in einsum regressed in performance improvements

### ISSUE THREAD

**Issue by t-vi** [pytorch/pytorch#58380]
## 🐛 Bug

From [Sohrab Samimi on the forums](https://discuss.pytorch.org/t/problem-with-einsum/121476):

```python
first = torch.rand(12,8192,2)
weights1 = torch.rand(12,8192,2)
torch.einsum('bix,iox->box',first,weights1)
```
gives

```
einsum() operands do not broadcast with remapped shapes [original->remapped]: [12, 8192, 2]->[12, 1, 2, 8192] [12, 8192, 2]->[1, 8192, 2, 12]
```

To which they say:
> Can someone explain to me what I am doing wrong?

Clearly, the `i`-dimensions don't match, but the error message sounds relatively opaque to me.

## Expected behavior

In my opinion, it would be better to print the letter/dimension and term number where the size mismatch happens. I think the old version did this, at least sometimes...

@heitorschueroff

**Comment by heitorschueroff** [comment]
@t-vi I agree that this error message can be improved. I went with a similar error message as NumPy but now looking back it's not easy to understand what is going on. I'll work on an improvement as part of the changes to einsum I'm doing. Thanks!

**Comment by t-vi** [comment]
@heitorschueroff Thanks for looking at this!

I really love your perf work on einsum, too!
